Orange/widgets/utils/itemmodels.py

```diff
@@ -63,7 +64,135 @@ def _as_contiguous_range(the_slice, length):
     return start, stop, step
 
 
-class PyTableModel(QAbstractTableModel):
+class AbstractSortTableModel(QAbstractTableModel):
+    """
+    A sorting proxy table model that sorts its rows in fast numpy,
+    avoiding potentially thousands of calls into
+    ``QSortFilterProxyModel.lessThan()`` or any potenrially costly
+    reordering of original data.
+
+    Override ``sortColumnData()``, adapting it to your underlying model.
+
+    Make sure to use ``mapToSource()``/``mapFromSource()`` whenever
+    fetching or manipulating table data, such as in ``data()``.
+
+    When updating the model (inserting, removing rows), the sort order
+    needs to be accounted for (e.g. reset and re-applied).
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.__sortInd = None     #: Indices sorting the source table
+        self.__sortIndInv = None  #: The inverse of __sortInd
+        self.__sortColumn = -1    #: Sort key column, or -1
+        self.__sortOrder = Qt.AscendingOrder
+
+    def sortColumnData(self, column):
+        """Return raw, sortable data for column"""
+        raise NotImplementedError
+
+    def _sortColumnData(self, column):
+        try:
+            # Call the overridden implementation if available
+            data = numpy.asarray(self.sortColumnData(column))
+        except NotImplementedError:
+            # Fallback to slow implementation
+            data = numpy.array([self.index(row, column).data()
+                                for row in range(self.rowCount())])
+
+        data = data[self.mapToSource(Ellipsis)]
+        return data
+
+    def sortColumn(self):
+        """The column currently used for sorting (-1 if no sorting is applied)"""
+        return self.__sortColumn
+
+    def sortOrder(self):
+        """The current sort order"""
+        return self.__sortOrder
+
+    def mapToSource(self, rows: int or numpy.ndarray or Ellipsis):
```

Comment:
This only evaluates to `def mapToSource(self, rows: int):`

`Union[int, Sequence[int], type(...)]` would be better.

However I would drop the Ellipsis unless you assure the returned array is is a copy or a read only view. Otherwise it makes little sense to keep `__sortInd` private.

I am also not too thrilled about the name. `mapToSourceRows` would be better (
I purposefully did not chose this name in TableModel in order to  differentiate from `QAbstractProxyModel.map{To,From}Source`)